### PR TITLE
internal/xdscache: only update ClusterLoadAssignment cache on changes

### LINKED ISSUE #3053: Only push ClusterLoadAssignments updates if they have changed
**What steps did you take and what happened:**

As reported on slack, after upgrading to 1.9 Contour is spamming the following log message:
```
level=info msg=stream_wait connection=1469 context=xds node_id=envoy-rklgl node_version=v1.15.1 resource_names="[default/<service_name>]" response_nonce=51581 type_url=type.googleapis.com/envoy.api.v2.ClusterLoadAssignment version_info=51581
```

This really should be a debug level log, but more serious is the rate of updates this indicates.

Most likely cause is the notification that is sent when the load assignment cache is rebuilt on each DAG run. The assumption
was that (1) The DAG notifications are less frequent than endpoint changes and (2) envoy would filter no-op updates with low cost.

We probably should revisit this and only notify if a DAG rebuild caused the cache to change.

### PR DESCRIPTION
**Pull request by skriss**
Only updates the ClusterLoadAssignment xDS cache and notifies via the
condition variable if the contents have changed during recalculation.

Closes #3053.

Signed-off-by: Steve Kriss <krisss@vmware.com>

### CODE REVIEW

**Comment by skriss** [comment]
This looked OK in my basic testing, but would appreciate any extra 👀  !

File: internal/xdscache/v2/endpointstranslator.go [hunk 1]
```diff
@@ -298,11 +298,38 @@ func (e *EndpointsTranslator) OnChange(d *dag.DAG) {
...
+
+		e.Notify()
+	} else {
+		e.Debug("cluster load assignments did not change")
+	}
+}
+
+// equal returns true if a and b are the same length, have the same set
+// of keys, and have proto-equivalent values for each key, or false otherwise.
+func equal(a, b map[string]*envoy_api_v2.ClusterLoadAssignment) bool {
```
Comment (stevesloka) [inline internal/xdscache/v2/endpointstranslator.go:317]: I was doing this with `equality.Semantic.DeepEqual` on my branch, but it wasn't working. I can test this out to see how it behaves.

File: internal/xdscache/v2/endpointstranslator.go [hunk 2]
```diff
@@ -298,11 +298,38 @@ func (e *EndpointsTranslator) OnChange(d *dag.DAG) {
...
+	if len(a) != len(b) {
+		return false
+	}
+
+	for k := range a {
+		if _, ok := b[k]; !ok {
```
Comment (stevesloka) [inline internal/xdscache/v2/endpointstranslator.go:323]: Does this assume ordering is the same each time?

File: internal/xdscache/v2/endpointstranslator.go (hunk 1 above)
Comment (skriss) [reply internal/xdscache/v2/endpointstranslator.go:317]: Yeah, I observed the same with that or `reflect.DeepEqual`, I think there are some private fields in there that don't match. Doing an item-by-item comparison using `proto.Equal` fixes it, though.

File: internal/xdscache/v2/endpointstranslator.go (hunk 2 above)
Comment (skriss) [reply internal/xdscache/v2/endpointstranslator.go:323]: Nope, no order assumptions. We're just checking each key from a to ensure it exists in b, and then comparing the values.

**Review by stevesloka — CHANGES_REQUESTED** [review]
Could you maybe add some tests to validate the `equal` method?

File: internal/xdscache/v2/endpointstranslator.go (hunk 1 above)
Comment (stevesloka) [reply internal/xdscache/v2/endpointstranslator.go:317]: Nice!

File: internal/xdscache/v2/endpointstranslator.go (hunk 2 above)
Comment (stevesloka) [reply internal/xdscache/v2/endpointstranslator.go:323]: 🎉

**Comment by skriss** [comment]
> Could you maybe add some tests to validate the `equal` method?

good call, added.

**Review by jpeach — APPROVED** [review]

**Pull request merged**